ggml : add repack testing support

danbev opened the pull request:
This PR consists of three commits but the first two are part of https://github.com/ggml-org/llama.cpp/pull/16004.

This commit add support for testing the ggml-cpu repack feature which
enables the repackaging of quantized data into more optimal layout for
matrix multiplication for specific hardware architectures.

The motivation is to enable the testing of a cpu backend that uses
repacked data against a reference cpu backend that does not use repacked
data.

Building:
```console
$ cmake -B build \
    -DGGML_CPU_REF_BACKEND=ON \
    -DGGML_BACKEND_DL=ON \
    -DGGML_CPU_ALL_VARIANTS=ON
```

List availble cpu architectures/variants:
```console
$ ./build/bin/test-backend-ops cpu-variants --list
CPU variants:
  CPU-haswell     - 12th Gen Intel(R) Core(TM) i7-1260P
  CPU-sse42       - 12th Gen Intel(R) Core(TM) i7-1260P
  CPU-x64         - 12th Gen Intel(R) Core(TM) i7-1260P
  CPU-alderlake   - 12th Gen Intel(R) Core(TM) i7-1260P
  CPU-sandybridge - 12th Gen Intel(R) Core(TM) i7-1260P

```
Run tests:
```console
./build/bin/test-backend-ops cpu-variants \
    --variant CPU-alderlake \
    -o "MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0,o=1)"

Testing CPU variant 'CPU-alderlake' against cpu-ref backend...

repack: repack tensor a with q4_0_8x8
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0,o=1): OK
repack: repack tensor a with q4_0_8x8
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0,o=1): OK
  14491/14491 tests passed
```
All matrix multiplication tests can be run by use specifying `-o "MUL_MAT, MUL_MAT_ID"`.

danbev committed:
ggml : add CPU backend reference implementation (wip)

This commit introduces a CPU reference implementation for GGML,
designed primarily for testing and validation purposes.

The motivation for this addition is to have a pure C CPU backend
implementation that does not use any hardware-specific optimizations
or intrinsics. This will allow for testing the CPU backend variants
against the reference implementation to ensure correctness

danbev committed:
move list_cpu_variants() to be called directly

danbev committed:
ggml : add repack testing support

This commit add support for testing the ggml-cpu repack feature which
enables the repackaging of quantized data into more optimal layout for
matrix multiplication for specific hardware architectures.

The motivation is to enable the testing of a cpu backend that uses
repacked data against a reference cpu backend that does not use repacked
data.

Building:
```console
$ cmake -B build \
    -DGGML_CPU_REF_BACKEND=ON
    -DGGML_BACKEND_DL=ON \
    -DGGML_CPU_ALL_VARIANTS=ON
```

List availble cpu architectures/variants:
```console
$ ./build/bin/test-backend-ops cpu-variants --list
CPU variants:
  CPU-alderlake   - 12th Gen Intel(R) Core(TM) i7-1260P
```
Run tests:
```console
./build-ref/bin/test-backend-ops cpu-variants \
    --variant CPU-alderlake \
    -o "MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0,o=1)"

Testing CPU variant 'CPU-alderlake' against cpu-ref backend...

repack: repack tensor a with q4_0_8x8
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0,o=1): OK
repack: repack tensor a with q4_0_8x8
  MUL_MAT(type_a=q4_0,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0,o=1): OK
  14491/14491 tests passed
```
All matrix multiplication tests can be run by use specifying
`-o "MUL_MAT"` but it may be harder to spot the ones that use repacking.

slaren commented on tests/CMakeLists.txt:
```diff
@@ -199,6 +199,15 @@ endif()
 llama_build_and_test(test-gguf.cpp)
 llama_build_and_test(test-backend-ops.cpp)
 
+if (GGML_CPU_REF_BACKEND)
+    if (WIN32)
+        set(GGML_CPU_REF_BACKEND_PATH "${CMAKE_BINARY_DIR}/bin/ggml-cpu-ref.dll")
+    else()
+        set(GGML_CPU_REF_BACKEND_PATH "${CMAKE_BINARY_DIR}/bin/libggml-cpu-ref.so")
+    endif()
+    target_compile_definitions(test-backend-ops PRIVATE GGML_CPU_REF_BACKEND_PATH="${GGML_CPU_REF_BACKEND_PATH}")
+endif()
```
As far as I can tell, `GGML_CPU_REF_BACKEND_PATH` is never used and this could be removed.

slaren reviewed (commented):
Strictly speaking, this is not enough to ensure that only the C implementation is used in the CPU-ref backend, because the base ABI may include some vector instructions. For example, the Linux AMD64 ABI includes SSE2, and if you add a feature flag for this instruction set, it will show up in the CPU-ref backend:

```diff
diff --git a/ggml/include/ggml-cpu.h b/ggml/include/ggml-cpu.h
index 9edd48513..004acea31 100644
--- a/ggml/include/ggml-cpu.h
+++ b/ggml/include/ggml-cpu.h
@@ -75,6 +75,7 @@ extern "C" {
     //

     // x86
+    GGML_BACKEND_API int ggml_cpu_has_sse2       (void);
     GGML_BACKEND_API int ggml_cpu_has_sse3       (void);
     GGML_BACKEND_API int ggml_cpu_has_ssse3      (void);
     GGML_BACKEND_API int ggml_cpu_has_avx        (void);
diff --git a/ggml/src/ggml-cpu/ggml-cpu.c b/ggml/src/ggml-cpu/ggml-cpu.c
index c13129084..06c8f54df 100644
--- a/ggml/src/ggml-cpu/ggml-cpu.c
+++ b/ggml/src/ggml-cpu/ggml-cpu.c
@@ -3428,6 +3428,14 @@ int ggml_cpu_has_llamafile(void) {
 #endif
 }

+int ggml_cpu_has_sse2(void) {
+#if defined(__SSE2__)
+    return 1;
+#else
+    return 0;
+#endif
+}
+
 int ggml_cpu_has_sse3(void) {
 #if defined(__SSE3__)
     return 1;
diff --git a/ggml/src/ggml-cpu/ggml-cpu.cpp b/ggml/src/ggml-cpu/ggml-cpu.cpp
index 90dd65c41..6f42b0a8f 100644
--- a/ggml/src/ggml-cpu/ggml-cpu.cpp
+++ b/ggml/src/ggml-cpu/ggml-cpu.cpp
@@ -506,6 +506,9 @@ static ggml_backend_feature * ggml_backend_cpu_get_features(ggml_backend_reg_t r
         ggml_cpu_init();

         std::vector<ggml_backend_feature> features;
+        if (ggml_cpu_has_sse2()) {
+            features.push_back({ "SSE2", "1" });
+        }
         if (ggml_cpu_has_sse3()) {
             features.push_back({ "SSE3", "1" });
         }
```

Other than that, I think this could be good enough for a first version. To be useful as part of a CI test however, `test-backend-ops` will need to be modified to be able to load every CPU backend variant supported by the system, not just the one selected by `ggml_backend_load_best`.

danbev committed:
ggml : add SSE2 detection for x86 backends

danbev commented:
> Other than that, I think this could be good enough for a first version.

Do you have any suggestions to how to also avoid compiler optimizations because of the availability of SSE2, or is it perhaps enough that the cpu-ref backend does not use any explicit SIMD operations?

danbev committed:
try workaround for windows

This commit adds ggml.c to the test-backend-ops compilation unit as on
windows the linker cannot resolve some of the functions that have been
moved into test-backend-ops, like ggml_hash_set_new. I'm not sure we
will keep this moving but for now this will at least allow the test to
be run on windows as well.

danbev committed:
ci : add workflow for testing CPU variants [no ci]

This is a work in progress to try out different CPU variants using the
CPU reference backend with repacking enabled. The workflow will be
manually triggered and allows specifying the operation and variant
to test.

danbev committed:
ggml : check src[1] does not have more than 2 dimensions

netrunnereve commented:
> Currently these problems go undetected and I think it would be useful to know how to apply the new CPU reference backend and the functionality in this PR to find these. Based on this, we can think if it would make sense to create some CI workflows to automate this (though we have to be mindful about the performance and how often to run those).

This actually can be done right now by doing a test-backend-ops against the Vulkan backend with llvmpipe. Obviously it would be way faster if we could directly compare two cpu implementations though.

danbev commented:
@slaren I've tried moving `ggml_backend_graph_copy` to `test-backend-ops.cpp` and I've got this to work. When you get a chance please take a look and see what you think. 

@ggerganov Regarding the testing I tried to use the cpu ref backend to find the SVE issue but it did not uncover it, but I did use the [test-cpu-variants](https://github.com/danbev/llama.cpp/actions/workflows/test-cpu-variants.yml) workflow for that issue which was useful just the same. This is currently a manual workflow with only a single job at this point. Just running the job without specifying a variant will list all the variants available, and a variant can be specified when running the workflow. It also supports disabling the repack backend as an example of options we could add. I'm not sure if you had something like this in mind for this but perhaps this can be a base for discussion?

I've kept the `tmate` action in the test-cpu-variants workflow for now which is an external action (non ggml-org), but I find it very useful to be able to ssh into the runner and run build/test manually.

ggerganov reviewed (commented):
> @ggerganov Regarding the testing I tried to use the cpu ref backend to find the SVE issue but it did not uncover it, but I did use the [test-cpu-variants](https://github.com/danbev/llama.cpp/actions/workflows/test-cpu-variants.yml) workflow for that issue which was useful just the same. This is currently a manual workflow with only a single job at this point. Just running the job without specifying a variant will list all the variants available, and a variant can be specified when running the workflow. It also supports disabling the repack backend as an example of options we could add. I'm not sure if you had something like this in mind for this but perhaps this can be a base for discussion?

I was thinking mainly about the following use case: if for example there is a bug in the ARM version of `ggml_vec_dot_q4_0_q8_0()` it would be nice to be able to detect it by comparing it to the `ggml_vec_dot_q4_0_q8_0_generic()` implementation.

slaren commented on ggml/src/ggml-cpu/repack.cpp:
```diff
@@ -1869,8 +1869,45 @@ static const ggml::cpu::tensor_traits * ggml_repack_get_optimal_repack_type(cons
+}
+
 static enum ggml_status ggml_backend_cpu_repack_buffer_init_tensor(ggml_backend_buffer_t buffer, struct ggml_tensor * tensor) {
-    tensor->extra = (void *) const_cast<ggml::cpu::tensor_traits *>(ggml_repack_get_optimal_repack_type(tensor));
+    if (tensor->op == GGML_OP_NONE) {
+        tensor->extra = (void *) const_cast<ggml::cpu::tensor_traits *>(ggml_repack_get_optimal_repack_type(tensor));
+        tensor->buffer = buffer;
+    }
+
+    if (supports_tensor(tensor)) {
+        tensor->src[0]->extra = (void *) const_cast<ggml::cpu::tensor_traits *>(ggml_repack_get_optimal_repack_type(tensor->src[0]));
+        tensor->src[0]->buffer = buffer;
+    }
```
This is definitely not ok, you can't change the buffer of a tensor during `init_tensor`, and you absolutely cannot change the buffer of a source tensor.
`test-backend-ops` needs to be modified to use the extra buffer types in the way they are intended - that is, only allocating the weights on the extra buffer type, and everything else on the default buffer type. This will likely mean either creating a new type of test, or making deep changes to the way tests are defined. Hacking the backend code to make it work is not ok.

danbev committed:
remove setting of buffers in init tensor [no ci]

danbev committed:
move repack hack to test-backend-ops.cpp [no ci]

This is just to get the tests passing again and I'll try to find a
better aproach to do this.

danbev committed:
repack : add check for 2D src[1] in supports_op

danbev committed:
tests : refactor handling of extra buffer types

danbev commented:
Closing this and will merge the non-repack related parts into #16004.
